File: backend/app/schemas/facial_metrics.py
from datetime import datetime
from pydantic import BaseModel, Field


class FacialFeatures(BaseModel):
    earMean: float | None = None
    earMin: float | None = None
    earStd: float | None = None
    marMean: float | None = None
    marMax: float | None = None
    marStd: float | None = None
    perclose: float | None = None
    eyesClosedRatio: float | None = None
    yawnCount: int | None = None


class FacialPrediction(BaseModel):
    modelName: str | None = None
    fatigueProbability: float | None = None
    fatigueDetected: bool | None = None
    focusScore: float | None = None


class FacialMetricsRequest(BaseModel):
    userId: str
    sessionId: str
    ear: float
    mar: float
    eyesClosed: bool
    yawning: bool
    timestamp: datetime | None = None
    features: FacialFeatures | None = None
    prediction: FacialPrediction | None = None
    inferenceTimeMs: float | None = None
    processingTimeMs: float | None = None


class FacialMetricsResponse(BaseModel):
    id: str = Field(..., alias="_id")
    sessionId: str
    ear: float
    mar: float
    eyesClosed: bool
    yawning: bool
    timestamp: datetime
    features: FacialFeatures | None = None
    prediction: FacialPrediction | None = None

    model_config = {
        "populate_by_name": True
    }

File: frontend/src/types/facialMetrics.ts
export type FacialMetrics = {
  _id: string;
  sessionId: string;
  ear: number;
  mar: number;
  eyesClosed: boolean;
  yawning: boolean;
  features?: FacialFeatures | null;
  prediction?: FacialPrediction | null;
  timestamp: string;
  inferenceTimeMs?: number | null;
  processingTimeMs?: number | null;
  requestTimestamp?: string | null;
};

export type FacialFeatures = {
  earMean?: number | null;
  earMin?: number | null;
  earStd?: number | null;
  marMean?: number | null;
  marMax?: number | null;
  marStd?: number | null;
  perclose?: number | null;
  eyesClosedRatio?: number | null;
  yawnCount?: number | null;
};

export type FacialPrediction = {
  modelName?: string | null;
  fatigueProbability?: number | null;
  fatigueDetected?: boolean | null;
  focusScore?: number | null;
};
